src/utils/tdex.ts

```diff
@@ -1,15 +1,167 @@
+      console.error(e);
+    }
+  }
+  return marketsByProvider;
+}
+
+/**
+ * Construct all the TradeOrder from a set of markets + a pair
+ * @param marketsByProvider the set of available markets mapped to provider endpoint
+ * @param pair the pair to get the orders for
+ */
+export function computeOrders(
+  pair: Pair,
+  marketsByProvider: Record<TDEXProvider['endpoint'], MarketInterface[]>,
+): TradeOrder[] {
+  const orders: TradeOrder[] = [];
+  for (const [providerEndpoint, markets] of Object.entries(marketsByProvider)) {
+    orders.push(...tdexOrdersForProvider(pair, providerEndpoint, markets));
+    if (orders.length > 0) {
+      return orders;
+    }
+  }
+
+  return orders;
+}
+
+// Create discoverer object for a specific set of trader clients
+function createDiscoverer(orders: TradeOrder[], errorHandler?: () => Promise<void>): Discoverer {
+  return new Discoverer(orders, combineDiscovery(bestBalanceDiscovery, bestPriceDiscovery), errorHandler);
+}
+
+/**
+ * discover the best provider for a pair
+ * @param orders orders to compare
+ * @param pair pair describing the trade
+ * @param type send or receive
+ * @param sats amount of `type` asset in satoshis
+ */
+export async function discoverBestOrder(orders: TradeOrder[], pair: Pair, type: keyof Pair, sats: number): Promise<TradeOrder> {
+  const discoverer = createDiscoverer(orders);
+  const best = await discoverer.discover({
+    asset: pair[type],
+    amount: sats
+  });
+
+  return best[0];
+}
+
+function createTrade(
+  order: TradeOrder,
+  utxos: UtxoInterface[],
+  explorerUrl: string
+): Trade {
+  return new Trade({
+    explorerUrl: explorerUrl,
+    providerUrl: order.traderClient.providerUrl,
+    utxos,
+    coinSelector: greedyCoinSelector(),
+  });
+}
+
+
+export async function calculateMarketPrice(
+  order: TradeOrder,
+  coinHash: string,
+  satoshis: number,
+) {
+  const [firstPrice] = await order.traderClient.marketPrice(
+    order.market,
+    order.type,
+    satoshis,
+    coinHash
+  );
+
+  return firstPrice.amount;
+}
+
+/**
+ * make a trade
+ * @param identity who sign/blind the tx
+ * @param known asset/amount in sats pair to trade
+ * @param order order describing the trade
+ * @param utxos unspents available to be selected for the trade
+ * @param explorerUrl the explorerURL (using to broadcast the tx)
+ */
+export async function makeTrade(
+  identity: IdentityInterface,
+  known: { sats: number, asset: string },
+  order: TradeOrder,
+  utxos: UtxoInterface[],
+  explorerUrl: string
+): Promise<string> {
+  try {
+    const trade = createTrade(order, utxos, explorerUrl);
+    const args: BuySellOpts = {
+      amount: known.sats,
+      asset: known.asset,
+      market: order.market,
+      identity
+    }
+    const promise = order.type === TradeType.BUY ? trade.buy(args) : trade.sell(args);
+    const txid = await promise;
+    return txid;
+  } catch (e) {
+    console.error(e);
+    throw new Error('error while making trade');
```

Comment:
Why not throw the `e.message` error itself?

Reply:
In fact, tdex-sdk does not return very explicit error messages (more about this [here](https://github.com/tdex-network/tdex-app/issues/443) and [here](https://github.com/tdex-network/tdex-sdk/issues/121)) that's why we wrap the buy/sell call with a try-catch statement in order to return a more "human-readable error msg" in tdex-app. 
However, looking at master, it looks like we just return the raw SDK errors so I've just dropped the try-catch statement and let `Trade.svelte` handle errors.